### ISSUE TITLE
[CVE-2021-37501] buffer overflow in h5dump 

### ISSUE THREAD

**Issue by e4t** [HDFGroup/hdf5#2458]
**Describe the bug**
[CVE-2021-37501](https://nvd.nist.gov/vuln/detail/CVE-2021-37501) (GHSA-rfgw-5vq3-wrjf) describes a buffer overflow in h5dump, a [reproducer](https://github.com/ST4RF4LL/Something_Found/blob/main/HDF5_v1.13.0_h5dump_heap_overflow.assets/poc) and [description](https://github.com/ST4RF4LL/Something_Found/blob/main/HDF5_v1.13.0_h5dump_heap_overflow.md) for this can be found in
https://github.com/ST4RF4LL/Something_Found.

The root cause is an artificially large size of a dimension in a dataspace message which causes the size of the on-disk data to exceed the addressable range and thus causes a multiplication to overflow. 
In `src/H5Oattr.c:H5O__attr_decode()`  with `ds_size` being set to 4611686018427387904 (0x4000000000000000) and `dt_size` equal to 8, the multiplication in:
```
H5_CHECKED_ASSIGN(attr->shared->data_size, size_t, ds_size * (hsize_t)dt_size, hsize_t);
```
overflows and is truncated to 0.
Since there is no provision to detect the overflow, the size is underestimated which will finally lead to a heap buffer overflow when the data is read with a subsequent segfault.

**Expected behavior**
The size overflow should be detected and the respective decode function should report an error so that the caller will terminate gracefully.

**Platform**
-  at least HDF5 1.10.8 and later (including current 'develop' branch)
-  openSUSE Leap 15.4/ SLE 15 SP4 / openSUSE Tumbleweed
- gcc7
- Autotools
- `configure --enable-fortran --enable-unsupported --enable-hl --enable-shared --enable-threadsafe --enable-build-mode=production --enable-cxx --with-pthread`

**Comment by Mingli-Yu** [comment]
Does the version 1.8.21 have the CVE-2021-37501 issue? Thanks!

**Comment by derobins** [comment]
> Does the version 1.8.21 have the [CVE-2021-37501](https://github.com/advisories/GHSA-rfgw-5vq3-wrjf) issue? Thanks!

Unfortunately, we no longer support HDF5 1.8

**Comment by Mingli-Yu** [comment]
> > Does the version 1.8.21 have the [CVE-2021-37501](https://github.com/advisories/GHSA-rfgw-5vq3-wrjf) issue? Thanks!
> 
> Unfortunately, we no longer support HDF5 1.8

Could you help hint if the CVE-2021-37501 issue exist in hdf5 1.8.21 or not?  Thanks!

**Comment by jhendersonHDF** [comment]
> > > Does the version 1.8.21 have the [CVE-2021-37501](https://github.com/advisories/GHSA-rfgw-5vq3-wrjf) issue? Thanks!
> > 
> > 
> > Unfortunately, we no longer support HDF5 1.8
> 
> Could you help hint if the [CVE-2021-37501](https://github.com/advisories/GHSA-rfgw-5vq3-wrjf) issue exist in hdf5 1.8.21 or not? Thanks!

@Mingli-Yu, yes, this CVE issue exists in the 1.8.21 release, as well as 1.8.22 and 1.8.23.

**Comment by mattjala** [comment]
The reproducer file can be found [here](https://github.com/HDFGroup/cve_hdf5/blob/main/cvefiles/cve-2021-37501.h5)

**Comment by lrknox** [comment]
Fixed by PR #2459 (develop).  Fix merged to hdf5_1_10 by PR #2579, to hdf5_1_12 by PR #2581 and to hdf5_1_14 by PR #2582.
